File: ENTEXAM.cpp
/*input
5
5 3 4 10
5 9 8 6
7 3 4 6
1 5 4 9
2 9 6 7
6 4 8
2 1 3 10
5 9 6
10 11
5 3 3 10
4 1 1 
1 1 1 
1 1 1 
1 1 1
1 0
4 2 3 10
7 7 7
4 6 10
7 10 9
9 9
5 4 3 10
0 0 0 
0 0 0
0 0 0 
0 0 0
0 0
1 1
*/
/*

1   30
2   40
3    21
4    28
5	20
*/
#include <iostream>
using namespace std;
#include <algorithm>
long long num[100000000];
bool cmpf(long long  a, long long b)
{
	return a > b;
}
int main()
{
	ios_base::sync_with_stdio(0);
	long long t;
	cin>>t;
	long long N,K,E,M,total,n;
	while(t--)
	{
		total=n=0;
		for(long long i=0;i<100000000;i++)
			num[i]=0;
		cin>>N>>K>>E>>M;
		for(long long i=0;i<N-1;i++)
		{
			for(long long j=0;j<E;j++)
			{
				cin>>n;
				num[i]+=n;
			}
		}
		/*for(long long j=0;j<N-1;j++)
		{
			cout<<num[j]<<" ";
		}*/
		sort(num,num+N+1,cmpf);
		/*cout<<endl;
		for(long long j=0;j<N-1;j++)
		{
			cout<<num[j]<<" ";
		}
		cout<<endl;*/
		for(long long j=0;j<E-1;j++)
		{
			cin>>n;
			total+=n;
		}
		//cout<<"total "<<total<<endl;
		if(total>num[K-1])
			cout<<"0";
		else {
			if(num[K-1]-total>=M)
				cout<<"Impossible";
			else 
				cout<<num[K-1]-total+1;
		}
		cout<<endl;
	}
}

File: ENTEXAM.java
/*
/* User_Id @deepakks
/* Problem_Code @ENTEXAM
/*
/*****************************************/


import java.util.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;



class TestClass
{
    public static void main(String args[] ) throws Exception
    {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        int N,K,E,M,min;
        long total,temp;
        String s[];
        while(T--!=0)
        {
            int check=0;
            s = br.readLine().split(" ");
            N = Integer.parseInt(s[0]);
            K = Integer.parseInt(s[1]);
            E = Integer.parseInt(s[2]);
            M = Integer.parseInt(s[3]);
            long num[] = new long [N-1];
            for(int i=0; i<N-1; i++)
            {
                s = br.readLine().split(" ");
                for(int j=0; j<E; j++)
                    num[i] += Integer.parseInt(s[0]);
            }
            Arrays.sort(num);/*
    		for(int i=0;i<N-1;i++)
    			System.out.print(num[i]+" ");*/
            for(int i=0; i<(N-1)/2; ++i)
            {
                temp = num[i];
                num[i] = num[N-1 - i - 1];
                num[N-1 - i -1] = temp;
            }/*
    		System.out.println();
    		for(int i=0;i<N-1;i++)
    			System.out.print(num[i]+" ");*/
            s = br.readLine().split(" ");
            total = 0;
            for(int i=0; i<E-1; i++)
                total += Integer.parseInt(s[i]);
            if(total > num[K-1] )
                System.out.println(0);
            else 
            {
                if(num[K-1]-total>=M)
                    System.out.println("Impossible");
                else System.out.println(num[K-1] - total + 1);
            }
        }
    }
}
